YOLOv3 for object detection (#4040)
mlpack/mlpack: doc/user/methods/yolov3.md

## YOLOv3 and YOLOv3Tiny

The `YOLOv3` and `YOLOv3Tiny` classes implement the models from
the paper ["YOLOv3: An Incremental Improvement"](https://arxiv.org/abs/1804.02767). `YOLOv3` is a simple
object detection algorithm that takes in a image and makes
thousands of guesses in a single neural network pass, which makes
these models really easy and quick to use.

**NOTE**: At the current time, only prediction is supported by the `YOLOv3`
and `YOLOv3Tiny` classes. Support for training and fine-tuning is in progress.

#### Simple usage example:

```c++
// Download: https://models.mlpack.org/yolo/yolov3-320.bin
// Predict bounding boxes from an image using `YOLOv3`
size_t imgSize = 320;
mlpack::YOLOv3 model;

> mlpack::Load("yolov3-320.bin", model);

Reviewer: Okay, so the "main" intended usage is to create an object and use `Load()` to load the weights.  That's totally fine but I think that we should make that clear in the constructors section (I'll point that out below).

#### Quick Links:

 * [Constructors](#constructors): create `YOLOv3` objects.
 * [`Predict()`](#predicting-bounding-boxes): predict bounding boxes in an
   image.
 * [Other functionality](#other-functionality) for loading, saving, and
   inspecting the model.
 * [`YOLOv3Tiny` class](#yolov3tiny) for efficient low-resource detection.
 * [Pretrained weights](#pretrained-weights) for different YOLO models.
 * [Examples](#simple-examples) of simple usage and links to detailed example
   projects.
 * [Template parameters](#advanced-functionality-template-parameters) for custom
   behavior.

#### See also:
 * [Object Detection on Wikipedia](https://en.wikipedia.org/wiki/Object_detection)
 * [You Only Look Once (original YOLO paper)](https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/Redmon_You_Only_Look_CVPR_2016_paper.pdf)
 * [YOLOv3 paper](https://arxiv.org/abs/1804.02767) where the YOLOv3 architecture is described.
 * [`DAGNetwork`](/src/mlpack/methods/ann/dag_network.hpp) is used internally to represent the model.

### Constructors

Construct a `YOLOv3` object using one of the constructors below.
Defaults and types are detailed in the

> [Constructor Parameters](#constructor-parameters) section below.

Reviewer: We don't have this table here, so we can remove this sentence.

 * `model = YOLOv3()`
   - Create an uninitialized YOLOv3 model.
   - After this step, load [pretrained weights](#pretrained-weights) into the
     model with [`Load()`](../load_save.md#load).

---

### Predicting Bounding Boxes

Once the weights are loaded, you can compute likely object bounding boxes with with `Predict()`.

#### Forms

<!-- TODO: could have one that takes input and output file names? so image Load/Save is abstracted too. -->

 * `model.Predict(image, opt, ignoreThresh)`
   - Predict objects in the given `image` (with metadata `opts`) and
     draw the detections onto that image.
   - Bounding boxes will not be drawn unless their confidence is
     greater than `ignoreThreshold`.

| **name** | **type** | **description** | **default**|
|----------|----------|-----------------|------------|
| `image` | `arma::fmat` | Input image. See [example](#simple-usage-example) for details on loading an image with mlpack | n/a |
| `opt` | `ImageOptions` | Image metadata. | n/a |
| `ignoreThreshold` | `double` | Minimum confidence to have the corresponding bounding box drawn onto `image`. | `0.7` |

---

 * `model.Predict(input, output)`

>   - Takes in a preprocessed `input`. See [example](#simple-examples)

Author: You'd also have to pass in the `ImageOptions`. I made a PreprocessImage method, I could just make that public?

  - Objectness means how likely an object is in the given box.
  - The class probability means that given there's an object in the box, what's the probability that it's this class.

| **name** | **type** | **description** |
|----------|----------|-----------------|
| `input` | `MatType` | Input image. See [example](#simple-usage-example) for details on preprocessing the input image |
| `output` | `MatType` | Raw outputs of the model |

---

### Other Functionality

 * `YOLOv3` can be serialized with
   [`Save()` and `Load()`](../load_save.md#mlpack-models-and-objects).

 * `Model()` will return the underlying
   [`DAGNetwork`](/src/mlpack/methods/ann/dag_network.hpp) object that represents the model architecture.

 * `ImageSize()` will return the width and height of the image after
    preprocessing. The `yolov3-320.bin` model for example converts the
    input image to shape `(320, 320, 3)`.

 * `NumAttributes()` will return the number of attributes that make up
    a bounding box prediction. For example, if the model predicts 80
    classes, it will return 85 to include the bounding box coordinates

>     and its objectness score.

Reviewer: Yes, definitely, the user needs some way of recovering what that list is.

 * `ClassNames()` will return a vector of strings, each being the name
    of a class the model can predict.

### `YOLOv3Tiny`

<!-- TODO: update when YOLOv3Tiny is updated. -->

### Pretrained weights

Because training a `YOLOv3` model from scratch is time-consuming,
a number of pretrained models are available for download:

The format for the name of each pretrained model is `<model name>-<image size>-<finetuned dataset name>.bin`.

 * [`https://models.mlpack.org/yolo/yolov3-320-coco.bin`](https://models.mlpack.org/yolov3/yolov3-320-coco.bin)

> <!-- TODO: update and add other models -->

Author: Well, theres different `YOLOv3`'s that were trained on different `imgSize`: 320, 416, 608. Also, wouldn't you rather the default be fmat? Does anyone do f64 neural networks? Also the default one would then be faster for the user if it was only f32. Like, I'd rather less options, but I can understand wanting to use arma::mat since that's what the rest of the docs do. Not sure.

Defaults and types are detailed in the

 * `model = YOLOv3()`
   - Create an uninitialized YOLOv3 model.
   - After this step, load [pretrained weights](#pretrained-weights) into the
     model with [`Load()`](../load_save.md#load).

### Predicting Bounding Boxes

Once the weights are loaded, you can compute likely object bounding boxes with with `Predict()`.

 * `model.Predict(image, opts, output, drawBoxes=false ignoreThresh=0.7)`
   - Predict objects in the given `image` (with metadata [`opts`](../load_save.md#imageoptions)).
   - Predict will apply a [letterbox](../core/images.md#letterbox-transform) transform to `image`. Image pixel values are expected to be between 0-255.
   - If `drawBoxes` is true, `output` will be a copy of image with the bounding boxes from the model drawn onto it.
   - Bounding boxes will be drawn if their confidence is greater than `ignoreThresh`.
   - If `drawBoxes` is false, the output will be the raw outputs of the model. The bounding box coordinates will be in the original image's space.

| **name** | **type** | **description** | **default**|
|----------|----------|-----------------|------------|

> | `image` | `arma::mat` | Input image. See [example](#simple-usage-example) for details on loading an image with mlpack | n/a |

Author: Should every mention link to matrices.md?

| `opt` | [`ImageOptions`](../load_save.md#imageoptions) | Image metadata. | n/a |
| `output` | `arma::mat` | Output. Will be an output image with bounding boxes or raw outputs of the model depending on `drawBoxes`. | n/a |
| `drawBoxes` | `bool` | Boolean to decide to copy `image` to `output` and draw bounding boxes or simply return raw outputs in `output`. | `false` |
| `ignoreThreshold` | `double` | Minimum confidence to have the corresponding bounding box drawn onto `output`, if `drawBoxes` is true. | `0.7` |

---

 * `model.Predict(preprocessedInput, rawOutput)`

>    - Takes in a `preprocessedInput`. See [example](#simple-examples).

Reviewer: ```suggestion
   - Takes in a manually preprocessed input image. See [example](#simple-examples).
```
This suggestion is just so we are super clear that it's manual and that the user has to do it.

>    - `rawOutput` stores the raw detection data. The shape of the output matrix will be `(numAttributes * numBoxes, batchSize)`.

Reviewer: ```suggestion
   - `rawOutput` stores the raw detection data. The shape of the output matrix will be `(numAttributes * numBoxes, preprocessedInput.n_cols)`.
```
(batch size is not defined anywhere)

   - Each bounding box is made up of `numAttributes` number of data points. This includes `cx`, `cy`, `w`, `h`, objectness and class probabilities.
   - You can get the `numAttributes` of the model from [`model.NumAttributes()`](#other-functionality).
   - The `numBoxes` depends on the model size. See the [pretrained weights](#pretrained-weights) section for more info.
   - You can get the `numBoxes` of the model from [`model.NumBoxes()`](#other-functionality).
   -  `cx` and `cy` are the coordinates for the center of the box. `w` and `h` are respectively the width and height of that bounding box.
   - Objectness means how likely an object is in the given box.

>    - The class probability means that given there's an object in the box, what's the probability that it's a particular class.

Reviewer: ```suggestion
   - Each bounding box is made up of `numAttributes` elements: `cx`, `cy`, `w`,
     `h`, the objectness score, and class probabilities for each class.
     * `cx` and `cy` are the coordinates for the center of the bounding box.
     * `w` and `h` are respectively the width and height of the bounding box.
     * Objectness represents how likely an object is in the given box (`0` to
       `1`).
     * The class probability represents the conditional probability that an
       object is of a particular class *given* that there is an object in the
       box.

   - `numAttributes` can be obtained from
     [`model.NumAttributes()`](#other-functionality).

   - `numBoxes` depends on the model size and can be obtained from
     [`model.NumBoxes()`](#other-functionality).  See the
     [pretrained weights](#pretrained-weights) section for more info.

   - So, e.g., `output(j * numAttributes + (5 + i), k)` obtains the class
     probability of the `i`th class in the `j`th bounding box of the `k`th image
     in `preprocessedInput`.  See [examples](#simple-examples) for usage
     samples.
```

Tell me what you think of this.  I refactored your text so that it uses sub-bullet points for organization, simplified the language just a little bit, then added an example.

Author: Yeah clearer.

of the `YOLOv3` class.

**NOTE**: You must define the `MLPACK_ENABLE_ANN_SERIALIZATION` macro
to serialize and deserialize models that use `arma::mat` as the data type.

Simple example loading the image, passing it to [`Predict()`](#predicting-bounding-boxes) and saving the output.

```c++
// Step 1: load the pretrained weights.
// Download: https://models.mlpack.org/yolo/yolov3-320-coco-f64.bin
mlpack::YOLOv3 model;
mlpack::Load("yolov3-320-coco-f64.bin", model);

// Step 2: load the image.
// Download: https://github.com/pjreddie/darknet/blob/master/data/dog.jpg
arma::mat inputImage, rawOutput;
mlpack::ImageOptions opts;
mlpack::Load("dog.jpg", inputImage, opts);

// Step 3: Preprocess the `inputImage`, predict bounding boxes using `YOLOv3`.
// Set `drawBoxes` to false in order to store raw outputs in `rawOutput`.
model.Predict(inputImage, opts, rawOutput, false);

// Step 4: Inspect the first possible bounding box.
std::cout << "First bounding box: [" << (size_t) rawOutput(0, 0) << ", "
    << (size_t) rawOutput(1, 0) << ", " << (size_t) rawOutput(2, 0) << ", "
    << (size_t) rawOutput(3, 0) << "]." << std::endl;
```

---

> Example of doing custom preprocessing on the input image, and getting raw output of the model.

Reviewer: ```suggestion
Example of doing manual preprocessing on the input image, and getting raw output of the model.
```

I feel like "manual" preprocessing is a better way to put it here?

// keeping the aspect ratio of the original image using `LetterboxImages`.
mlpack::ImageOptions preprocessedOpts = opts;
const size_t imgSize = model.ImageSize();
const double greyValue = 0.5;
LetterboxImages(preprocessedImage, preprocessedOpts, imgSize, imgSize, greyValue);

// Change the layout of the channels such that they're grouped.
preprocessedImage = GroupChannels(preprocessedImage, preprocessedOpts);

// Step 4: detect objects in the image.
// Get raw output from model and store in `rawOutput`.
model.Predict(preprocessedImage, rawOutput);

// Step 5: Inspect the first possible bounding box.
std::cout << "First bounding box: [" << (size_t) rawOutput(0, 0) << ", "
    << (size_t) rawOutput(1, 0) << ", " << (size_t) rawOutput(2, 0) << ", "
    << (size_t) rawOutput(3, 0) << "]." << std::endl;
```

---

Example of predicting and drawing with multiple images simultaneously. **NOTE**: On this example, each image must have the same dimensions.

```c++
// Step 1: load the pretrained model.
// Download: https://models.mlpack.org/yolo/yolov3-320-coco-f64.bin
mlpack::YOLOv3 model;
mlpack::Load("yolov3-320-coco-f64.bin", model);

> // Step 2: load the images, in this case copies of the same image three times.

Reviewer: ```suggestion
// Step 2: load the images.
```

## YOLOv3 and YOLOv3Tiny

The `YOLOv3` and `YOLOv3Tiny` classes implement the models from
the paper ["YOLOv3: An Incremental Improvement"](https://arxiv.org/abs/1804.02767). `YOLOv3` is a simple
object detection algorithm that takes in an image and predicts multiple bounding
boxes in a single forward pass of the neural network.

**NOTE**: At the current time, only prediction is supported by the `YOLOv3`
and `YOLOv3Tiny` classes. Support for training and fine-tuning is in progress.

#### Simple usage example:

**NOTE**: You must define the `MLPACK_ENABLE_ANN_SERIALIZATION` macro
to load a `YOLOv3` model from disk.

```c++
// Download: https://models.mlpack.org/yolo/yolov3-320-coco-f64.bin
mlpack::YOLOv3 model;
mlpack::Load("yolov3-320-coco-f64.bin", model);

// Download: https://models.mlpack.org/yolo/dog.jpg
arma::mat inputImage, outputImage;
mlpack::ImageOptions opts;
mlpack::Load("dog.jpg", inputImage, opts);

> // Preprocess the `inputImage`, predict bounding boxes using `YOLOv3` and draw them onto `outputImage`.

Reviewer: ```suggestion
// Preprocess the `inputImage`, predict bounding boxes using `YOLOv3` and draw
// them onto `outputImage`.
```
I try to keep this <= 80 characters wide so they aren't too wide in the browser (mobile or desktop).

     model with [`Load()`](../load_save.md#load).

### Predicting Bounding Boxes

Once the weights are loaded, you can compute likely object bounding boxes with with `Predict()`.

 * `model.Predict(image, opts, output, drawBoxes=false ignoreThresh=0.7)`
   - Predict objects in the given `image` (with metadata [`opts`](../load_save.md#imageoptions)).
   - Predict will apply a [letterbox](../core/images.md#letterbox-transform) transform to `image`. Image pixel values are expected to be between 0-255.
   - If `drawBoxes` is true, `output` will be a copy of image with the bounding boxes from the model drawn onto it.
   - Bounding boxes will be drawn if their confidence is greater than `ignoreThresh`.
   - If `drawBoxes` is false, the output will be the raw outputs of the model. The bounding box coordinates will be in the original image's space.
     * The format of the `output` matrix is described in the documentation for the next overload, below.

| **name** | **type** | **description** | **default**|
|----------|----------|-----------------|------------|
| `image` | [`arma::mat`](../matrices.md) | Input [image](../load_save.md#image-data). | n/a |

> | `opt` | [`ImageOptions`](../load_save.md#imageoptions) | Image metadata. | n/a |

Reviewer: ```suggestion
| `opts` | [`ImageOptions`](../load_save.md#imageoptions) | Image metadata. | n/a |
```

    bounding box could contain. For example, the pretrained weights
    were trained on the COCO dataset, which includes 80 different
    classes.

 * `ClassNames()` will return a vector of strings, each being the name
    of a class the model can predict. You can find all the COCO names in order [here](https://models.mlpack.org/yolo/coco.names).

 * `Anchors()` returns a vector of doubles representing the anchors used
    during inference.

 * `NumBoxes()` returns the number of possible bounding boxes the model can detect.

### `YOLOv3Tiny`

`YOLOv3Tiny` is a smaller object detection model based off of `YOLOv3` but for
low-resource machines. `YOLOv3Tiny` has an identical API to `YOLOv3` so it can
be used as a drop-in replacement. `YOLOv3` has ~60 million parameters, while
`YOLOv3Tiny` has ~8 million.

[Pretrained weights](#pretrained-weights) are also included for `YOLOv3Tiny`.

#### Simple example usage of `YOLOv3Tiny`

```c++
// Step 1: load the pretrained `YOLOv3Tiny` weights.
// Download: https://models.mlpack.org/yolo/yolov3-tiny-416-coco-f64.bin
mlpack::YOLOv3Tiny model;
mlpack::Load("yolov3-tiny-416-coco-f64.bin", model);

> // Step 2: load the image

Reviewer: ```suggestion
// Step 2: load the image.
```

mlpack::ImageOptions opts;
mlpack::Load("dog.jpg", image, opts);

// Step 3: Preprocess the input image, detect bounding boxes and draw them onto `outputImage`.
model.Predict(image, opts, outputImage, true);

// Step 4: Save to "output.jpg"
mlpack::Save("output.jpg", outputImage, opts, true);
```

### Pretrained weights

Because training a `YOLOv3` model from scratch is time-consuming,
a number of pretrained models are available for download.

The pretrained weights were trained on the COCO dataset.

 * [COCO dataset](https://cocodataset.org/) was used to train the pretrained weights included.
 * [COCO class names](https://models.mlpack.org/yolo/coco.names) in their correct order.

The format for the name of each YOLOv3 pretrained model is
`<model name>-<image size>-<finetuned dataset name>-<matrix type>.bin`.

 * [`yolov3-320-coco-f64.bin`](https://models.mlpack.org/yolov3/yolov3-320-coco-f64.bin)
 * [`yolov3-416-coco-f64.bin`](https://models.mlpack.org/yolov3/yolov3-416-coco-f64.bin)
 * [`yolov3-608-coco-f64.bin`](https://models.mlpack.org/yolov3/yolov3-608-coco-f64.bin)
 * [`yolov3-tiny-416-coco-f64.bin`](https://models.mlpack.org/yolov3/yolov3-tiny-416-coco-f64.bin)

> `arma::fmat` weights are also included. These weights [need custom template behaviour](#advanced-functionality-template-parameters).

Reviewer: ```suggestion
`arma::fmat` weights (e.g. 32-bit precision) are also available. These weights [need custom template behaviour](#advanced-functionality-template-parameters).
```

## YOLOv3 and YOLOv3-tiny

The `YOLOv3` and `YOLOv3-tiny` classes implement the models from

> the paper "YOLOv3: An Incremental Improvement". `YOLOv3` is a simple

Reviewer: It would probably be worth adding a link to the paper here, as well as a 'see also' section (like the other methods) that also links to the right place.

object detection algorithm that takes in a square image and makes
thousands of guesses in a single neural network pass, which makes
these models really easy and quick to use.

Both the YOLOv3 and YOLOv3-tiny classes have identical APIs.

> **NOTE**: Currently, training is being implemented and not ready yet.

Reviewer: ```suggestion
**NOTE**: At the current time, only prediction is supported by the `YOLOv3`
class.  Support for training and fine-tuning is in progress.
```
To me a natural question after reading the phrase is "ok, well, what *is* implemented"---so this suggestion pre-empts the question.

#### Simple usage example:

```cpp
  mlpack::YOLOv3<mlpack::EmptyLoss, mlpack::RandomInitialization, arma::fmat> model;

>   mlpack::Load(modelFile, model);

Reviewer: Can you add comments in the style of the rest of the documentation to this usage example?  Many users will only ever glance at this and use it as a starting point to adapt their documentation.

In addition adding a URL to where the model is is really important---the documentation snippet won't build until you do that, and this also won't be a fully standalone example.
